File: sdks/python/src/devhub_sdk/runtime.py
from __future__ import annotations

import os
import platform
from abc import ABC, abstractmethod
from pathlib import Path

from ._json import load_json_text
from ._parsing import parse_hub_runtime, validate_runtime_endpoints
from .models import DevHubClientOptions, RuntimeConnectionInfo


DATA_DIR_ENV = "DEVHUB_DATA_DIR"


class RuntimeResolver(ABC):
    """运行时发现抽象。"""

    @abstractmethod
    def resolve(self, options: DevHubClientOptions) -> RuntimeConnectionInfo:
        """根据数据根目录解析 Hub 连接信息。"""


class FileSystemRuntimeResolver(RuntimeResolver):
    """默认的文件系统运行时发现实现。"""

    def resolve(self, options: DevHubClientOptions) -> RuntimeConnectionInfo:
        cloned = options.clone()
        cloned.validate()

        data_directory = resolve_data_directory(cloned.data_dir)
        runtime_directory = data_directory / "runtime"
        hub_json_path = runtime_directory / "hub.json"
        if not hub_json_path.is_file():
            _raise_invalid_data_directory_error_if_needed(data_directory)
            raise RuntimeError(f"未找到 hub.json：{hub_json_path}")

        runtime_payload = load_json_text(hub_json_path.read_text(encoding="utf-8"), source=str(hub_json_path))
        runtime = parse_hub_runtime(runtime_payload, source=str(hub_json_path))

        token_path = Path(runtime.token_file)
        if not token_path.is_file():
            raise RuntimeError(f"未找到 token 文件：{token_path}")

        token = token_path.read_text(encoding="utf-8").strip()
        if not token:
            raise RuntimeError(f"token 文件为空：{token_path}")

        return RuntimeConnectionInfo(
            runtime_directory=str(runtime_directory),
            token=token,
            runtime=runtime,
        )


def discover_runtime(options: DevHubClientOptions) -> RuntimeConnectionInfo:
    """根据数据根目录发现 Hub 连接信息。"""

    return FileSystemRuntimeResolver().resolve(options)


def validate_runtime_connection_info(connection_info: RuntimeConnectionInfo, source: str = "runtime_resolver") -> None:
    """校验注入式运行时连接信息仍满足协议端点约束。"""

    validate_runtime_endpoints(connection_info.runtime.http_base_url, connection_info.runtime.ws_url, source)

    if connection_info.rpc_endpoint != f"{connection_info.runtime.http_base_url}/rpc":
        raise RuntimeError(f"{source}.rpc_endpoint 非法。")
    if connection_info.websocket_endpoint != connection_info.runtime.ws_url:
        raise RuntimeError(f"{source}.websocket_endpoint 非法。")


def resolve_data_directory(data_dir_override: str | None = None) -> Path:
    """解析运行时数据根目录。"""

    if data_dir_override and data_dir_override.strip():
        return _to_absolute_path(Path(data_dir_override).expanduser())

    env_data_dir = os.getenv(DATA_DIR_ENV)
    if env_data_dir and env_data_dir.strip():
        return _to_absolute_path(Path(env_data_dir).expanduser())

    system = platform.system()
    home = Path.home()
    if system == "Windows":
        local_app_data = os.getenv("LOCALAPPDATA")
        base = Path(local_app_data) if local_app_data else home / "AppData" / "Local"
        return _to_absolute_path(base / "DevHub")
    if system == "Darwin":
        return _to_absolute_path(home / "Library" / "Application Support" / "DevHub")

    xdg_data_home = os.getenv("XDG_DATA_HOME")
    base = Path(xdg_data_home).expanduser() if xdg_data_home else home / ".local" / "share"
    return _to_absolute_path(base / "DevHub")


def _to_absolute_path(path: Path) -> Path:
    return Path(os.path.abspath(path))


def _raise_invalid_data_directory_error_if_needed(data_directory: Path) -> None:
    direct_hub_json_path = data_directory / "hub.json"
    direct_token_path = data_directory / "token.txt"
    if direct_hub_json_path.is_file() or direct_token_path.is_file():
        raise RuntimeError(
            "data_dir 必须指向数据根目录，SDK 仅支持 <data_dir>/runtime/hub.json；"
            f"不再支持直接传入 runtime 子目录或旧版 hub.json 直放布局：{data_directory}"
        )

File: sdks/javascript/src/runtime.ts
import { promises as fs } from "node:fs";
import os from "node:os";
import path from "node:path";
import type { NormalizedDevHubClientOptions } from "./models.js";
import { parseDateTimeString } from "./validation.js";

export const DATA_DIR_ENV = "DEVHUB_DATA_DIR";

export interface HubRuntimeTuning {
  leaseSeconds: number;
  onlineThresholdSeconds: number;
  launchDedupeWindowSeconds: number;
}

export interface HubRuntime {
  protocolVersion: number;
  pid: number;
  httpBaseUrl: string;
  wsUrl: string;
  tokenFile: string;
  startedAtUtc: Date;
  runtimeTuning: HubRuntimeTuning;
  hubVersion?: string;
}

export interface RuntimeConnectionInfo {
  runtimeDirectory: string;
  token: string;
  runtime: HubRuntime;
  rpcEndpoint: string;
  websocketEndpoint: string;
}

export interface RuntimeResolver {
  resolve(options: Readonly<NormalizedDevHubClientOptions>): Promise<RuntimeConnectionInfo>;
}

export class FileSystemRuntimeResolver implements RuntimeResolver {
  async resolve(options: Readonly<NormalizedDevHubClientOptions>): Promise<RuntimeConnectionInfo> {
    return discoverRuntime(options.dataDir);
  }
}

export function resolveDataDirectory(dataDirOverride?: string): string {
  if (dataDirOverride && dataDirOverride.trim()) {
    return path.resolve(dataDirOverride);
  }

  const envDataDir = process.env[DATA_DIR_ENV];
  if (envDataDir && envDataDir.trim()) {
    return path.resolve(envDataDir);
  }

  const platform = process.platform;
  const home = os.homedir();

  if (platform === "win32") {
    const base = process.env.LOCALAPPDATA ?? path.join(home, "AppData", "Local");
    return path.resolve(base, "DevHub");
  }

  if (platform === "darwin") {
    return path.resolve(home, "Library", "Application Support", "DevHub");
  }

  const xdgDataHome = process.env.XDG_DATA_HOME;
  const base = xdgDataHome && xdgDataHome.trim() ? xdgDataHome : path.join(home, ".local", "share");
  return path.resolve(base, "DevHub");
}

export async function discoverRuntime(dataDirOverride?: string): Promise<RuntimeConnectionInfo> {
  const dataDirectory = resolveDataDirectory(dataDirOverride);
  const { runtimeDirectory, hubJsonPath } = await resolveHubRuntimePaths(dataDirectory);

  let hubJsonText: string;
  try {
    hubJsonText = await fs.readFile(hubJsonPath, "utf-8");
  } catch {
    throw new Error(`未找到 hub.json：${hubJsonPath}`);
  }

  let payload: unknown;
  try {
    payload = JSON.parse(hubJsonText) as unknown;
  } catch {
    throw new Error(`hub.json 解析失败：${hubJsonPath}`);
  }

  const runtime = parseHubRuntime(payload, hubJsonPath);

  if (!path.isAbsolute(runtime.tokenFile)) {
    throw new Error(`hub.json.tokenFile 非法：${hubJsonPath}`);
  }

  let tokenText: string;
  try {
    tokenText = await fs.readFile(runtime.tokenFile, "utf-8");
  } catch {
    throw new Error(`未找到 token 文件：${runtime.tokenFile}`);
  }

  const token = tokenText.trim();
  if (!token) {
    throw new Error(`token 文件为空：${runtime.tokenFile}`);
  }

  return {
    runtimeDirectory,
    token,
    runtime,
    rpcEndpoint: `${runtime.httpBaseUrl}/rpc`,
    websocketEndpoint: runtime.wsUrl
  };
}

async function resolveHubRuntimePaths(dataDirectory: string): Promise<{
  runtimeDirectory: string;
  hubJsonPath: string;
}> {
  const legacyHubJsonPath = path.join(dataDirectory, "hub.json");
  const legacyTokenFilePath = path.join(dataDirectory, "token.txt");
  const [hasLegacyHubJson, hasLegacyTokenFile] = await Promise.all([
    fileExists(legacyHubJsonPath),
    fileExists(legacyTokenFilePath)
  ]);

  if (hasLegacyHubJson || hasLegacyTokenFile) {
    throw new Error(
      `dataDir 必须指向数据根目录，不支持直接传入 runtime 子目录或旧版 hub.json/token.txt 直放布局：${dataDirectory}`
    );
  }

  return {
    runtimeDirectory: path.join(dataDirectory, "runtime"),
    hubJsonPath: path.join(dataDirectory, "runtime", "hub.json")
  };
}

function parseHubRuntime(payload: unknown, source: string): HubRuntime {
  if (!isRecord(payload)) {
    throw new Error(`hub.json 解析失败：${source}`);
  }

  const protocolVersion = readNumber(payload, "protocolVersion", source);
  if (protocolVersion !== 1) {
    throw new Error(`hub.json.protocolVersion 非法：${source}`);
  }

  const pid = readInteger(payload, "pid", source);
  if (pid < 1) {
    throw new Error(`hub.json.pid 非法：${source}`);
  }

  const httpBaseUrl = readString(payload, "httpBaseUrl", source);
  validateHttpBaseUrl(httpBaseUrl, source);

  const wsUrl = readString(payload, "wsUrl", source);
  validateWebSocketUrl(wsUrl, source);

  const tokenFile = readString(payload, "tokenFile", source);

  const startedAtUtcRaw = readString(payload, "startedAtUtc", source);
  const startedAtUtc = parseDateTimeString(startedAtUtcRaw, "hub.json.startedAtUtc");

  const runtimeTuning = parseRuntimeTuning(payload.runtimeTuning, source);

  const hubVersion = readOptionalString(payload, "hubVersion", source);

  return {
    protocolVersion,
    pid,
    httpBaseUrl,
    wsUrl,
    tokenFile,
    startedAtUtc,
    runtimeTuning,
    hubVersion
  };
}

function parseRuntimeTuning(payload: unknown, source: string): HubRuntimeTuning {
  if (!isRecord(payload)) {
    throw new Error(`hub.json.runtimeTuning 非法：${source}`);
  }

  const leaseSeconds = readInteger(payload, "leaseSeconds", source, "hub.json.runtimeTuning 非法");
  const onlineThresholdSeconds = readInteger(payload, "onlineThresholdSeconds", source, "hub.json.runtimeTuning 非法");
  const launchDedupeWindowSeconds = readInteger(payload, "launchDedupeWindowSeconds", source, "hub.json.runtimeTuning 非法");

  if (leaseSeconds < 1 || onlineThresholdSeconds < 1 || launchDedupeWindowSeconds < 1) {
    throw new Error(`hub.json.runtimeTuning 非法：${source}`);
  }

  return {
    leaseSeconds,
    onlineThresholdSeconds,
    launchDedupeWindowSeconds
  };
}

function validateHttpBaseUrl(value: string, source: string): void {
  if (!value || value.endsWith("/")) {
    throw new Error(`hub.json.httpBaseUrl 非法：${source}`);
  }

  let url: URL;
  try {
    url = new URL(value);
  } catch {
    throw new Error(`hub.json.httpBaseUrl 非法：${source}`);
  }

  if ((url.protocol !== "http:" && url.protocol !== "https:") || !isLoopbackHost(url.hostname)) {
    throw new Error(`hub.json.httpBaseUrl 非法：${source}`);
  }

  if (url.username || url.password || url.pathname !== "/" || url.search || url.hash) {
    throw new Error(`hub.json.httpBaseUrl 非法：${source}`);
  }
}

function validateWebSocketUrl(value: string, source: string): void {
  if (!value || value.endsWith("/")) {
    throw new Error(`hub.json.wsUrl 非法：${source}`);
  }

  let url: URL;
  try {
    url = new URL(value);
  } catch {
    throw new Error(`hub.json.wsUrl 非法：${source}`);
  }

  if ((url.protocol !== "ws:" && url.protocol !== "wss:") || !isLoopbackHost(url.hostname)) {
    throw new Error(`hub.json.wsUrl 非法：${source}`);
  }
}

function isLoopbackHost(host: string): boolean {
  const normalized = host.replace(/^\[(.*)\]$/, "$1").toLowerCase();
  return normalized === "localhost" || normalized === "127.0.0.1" || normalized === "::1";
}

async function fileExists(target: string): Promise<boolean> {
  try {
    await fs.access(target);
    return true;
  } catch {
    return false;
  }
}

function readString(payload: Record<string, unknown>, key: string, source: string): string {
  const value = payload[key];
  if (typeof value !== "string" || !value.trim()) {
    throw new Error(`hub.json.${key} 非法：${source}`);
  }
  return value;
}

function readOptionalString(payload: Record<string, unknown>, key: string, source: string): string | undefined {
  if (!(key in payload)) {
    return undefined;
  }

  const value = payload[key];
  if (value === undefined) {
    return undefined;
  }

  if (typeof value !== "string") {
    throw new Error(`hub.json.${key} 非法：${source}`);
  }

  return value;
}

function readNumber(payload: Record<string, unknown>, key: string, source: string, overrideMessage?: string): number {
  const value = payload[key];
  if (typeof value !== "number" || Number.isNaN(value)) {
    const message = overrideMessage ? `${overrideMessage}：${source}` : `hub.json.${key} 非法：${source}`;
    throw new Error(message);
  }
  return value;
}

function readInteger(payload: Record<string, unknown>, key: string, source: string, overrideMessage?: string): number {
  const value = readNumber(payload, key, source, overrideMessage);
  if (!Number.isInteger(value)) {
    const message = overrideMessage ? `${overrideMessage}：${source}` : `hub.json.${key} 非法：${source}`;
    throw new Error(message);
  }

  return value;
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}
